### PR TITLE
feat(process): add cli parameters for importing directly to sphere

### PR DESCRIPTION
**Pull request by junajan**

### CODE REVIEW

**Comment by junajan** [comment]
Integration tests don't work because of missing sphere.io credentials in travis env variables.

File: src/coffee/product-type-import.coffee [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      options.protocol = argv.sphereProtocol if argv.sphereProtocol
+      if argv.sphereAuthHost
+        options.oauth_host = argv.sphereAuthHost
+        options.rejectUnauthorized = false
+      options.oauth_protocol = argv.sphereAuthProtocol if argv.sphereAuthProtocol
+
+      config =
+        sphereClientConfig: options
+
+      Promise.resolve config
```
Comment (PhilippSpo) [inline src/coffee/product-type-import.coffee]: No need for `Promise.resolve`. You can just return config here since you are already in the `then` callback.
Also in line `49`

File: src/coffee/product-type-import.coffee [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+  ###
+  Create sphere product import class
+  ###
+  _ensureProductTypeImporter: (argv, logger) ->
+    @_ensureCredentials argv
+    .then (credentials) ->
+      new ProductTypeImport.default(logger, credentials)
```
Comment (PhilippSpo) [inline src/coffee/product-type-import.coffee]: In coffescript there is also a "fat"-arrow
so you can simply write:
```coffeescript
_ensureProductTypeImporter: (argv, logger) ->
     @_ensureCredentials argv
     .then (credentials) =>
       @sphereImporter = new ProductTypeImport.default(logger, credentials)
```

File: src/spec/product-type-importer.spec.coffee [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      done(e)
+
+  it 'should not import wrong product type', (done) ->
+    expect(importer).toBeDefined()
+    expect(sphereClient).toBeDefined()
+
+    delete testProductType.name
+    importer.import {productTypes: [testProductType]}
+    .then ->
+      done "Importer wrong product type"
```
Comment (PhilippSpo) [inline src/spec/product-type-importer.spec.coffee]: how does this test that the "wrong" product type has not been importer?

Comment (junajan) [inline src/spec/product-type-importer.spec.coffee]: if it won't raise an error it will take this test as failed .. I will add there some check against sphere

File: src/coffee/run.coffee [hunk 4]
```diff
@@ -54,6 +75,50 @@ zipFiles = (path, filename) ->
       buffer = zip.generate type: 'nodebuffer'
       fs.writeFileAsync "#{path}/#{filename}.zip", buffer, 'utf8'
 
+saveProductTypes = (result) ->
+  console.log 'About to write files...'
```
Comment (butenkor) [inline src/coffee/run.coffee]: why console and not logger?

Comment (junajan) [inline src/coffee/run.coffee]: there is used console everywhere so this was to keep consistency

File: src/spec/product-type-importer.spec.coffee (hunk 3 above)
Comment (junajan) [inline src/spec/product-type-importer.spec.coffee]: updated 6c34f70

**Comment by PhilippSpo** [comment]
@junajan can you fix the breaking test also? Would be awesome 😊

File: .travis.yml [hunk 5]
```diff
@@ -1,28 +1,32 @@
 language: node_js
 node_js:
-  - '0.10'
-  - '0.12'
+- '0.10'
```
Comment (emmenko) [inline .travis.yml]: I think we can get rid of this. And we should test on `4` as well

File: .travis.yml [hunk 6]
```diff
@@ -1,28 +1,32 @@
...
 script:
-  - npm test
-  - cucumber -t ~@wip
-  - 'mkdir -p tmp && ./bin/product-type-generator --types data/sample-product-types.csv --attributes data/sample-product-types-attributes.csv --target ./tmp --zip'
-  - 'mkdir -p tmp && ./bin/product-type-generator --types data/marketplace-types.csv --attributes data/marketplace-attributes.csv --target ./tmp --withRetailer --zip'
-  - './bin/product-type-update --projectKey ${SPHERE_PROJECT_KEY} --clientId ${SPHERE_CLIENT_ID} --clientSecret ${SPHERE_CLIENT_SECRET} --source ./tmp'
+- npm test
+- cucumber -t ~@wip
+- mkdir -p tmp && ./bin/product-type-generator --types data/sample-product-types.csv
+  --attributes data/sample-product-types-attributes.csv --target ./tmp --zip
```
Comment (emmenko) [inline .travis.yml]: Why is this in a new line?

File: features/common.feature [hunk 7]
```diff
@@ -39,10 +39,10 @@ Feature: Show common information for tooling
           "attributeConstraint": "CombinationUnique",
           "isRequired": false,
           "isSearchable": false,
-          "inputHint": ""
-        }
+          "inputHint": "SingleLine"
+        },
     """
-    When I run `product-type-update --projectKey import-101-64 --source product-type-boo-txt-num.json`
+    When I run `product-type-update --projectKey nicola --source product-type-boo-txt-num.json`
```
Comment (emmenko) [inline features/common.feature]: Why is this changed?

File: package.json [hunk 8]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "sphere-product-type-json-generator",
   "description": "A command line tool for generating ProductTypes JSON representations from CSV templates.",
-  "version": "0.2.8",
+  "version": "0.3.0",
```
Comment (emmenko) [inline package.json]: Don't bump the version now. This is usually done automatically by the release task.

File: package.json [hunk 9]
```diff
@@ -40,7 +40,8 @@
     "node": ">= 0.10.0"
```
Comment (emmenko) [inline package.json]: We can require node 4 here.

File: src/coffee/product-type-import.coffee [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+{ExtendedLogger, ProjectCredentialsConfig} = require 'sphere-node-utils'
+package_json = require '../package.json'
+
+class ProductTypeImporter
+
+  ###
+  Prepare for importing, initialize used modules
+  ###
+  init: (argv) ->
+    @argv = argv
```
Comment (emmenko) [inline src/coffee/product-type-import.coffee]: Is this necessary?

File: src/coffee/product-type-import.coffee [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    @logOptions =
+      name: "#{package_json.name}-#{package_json.version}"
+      silent: !! argv.logSilent
+      streams: [
+        { level: 'error', stream: process.stderr }
+        { level: argv.logLevel || 'info', path: (argv.logDir || '.') + "/#{package_json.name}.log" }
```
Comment (emmenko) [inline src/coffee/product-type-import.coffee]: Nitpick: `path: "#{argv.logDir || '.'}/#{package_json.name}.log"`

File: src/coffee/run.coffee [hunk 12]
```diff
@@ -21,11 +42,11 @@ argv = require('optimist')
   .alias('target', 'td')
   .alias('withRetailer', 'r')
   .boolean('withRetailer')
+  .boolean('sphereImport')
```
Comment (emmenko) [inline src/coffee/run.coffee]: I would simply name it `import`

File: src/coffee/product-type-import.coffee [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        options.oauth_host = argv.sphereAuthHost
+        options.rejectUnauthorized = false
+      options.oauth_protocol = argv.sphereAuthProtocol if argv.sphereAuthProtocol
+
+      sphereClientConfig: options
+
+  ###
+  Create sphere product import class
+  ###
+  _ensureProductTypeImporter: (argv, logger) ->
```
Comment (emmenko) [inline src/coffee/product-type-import.coffee]: This function doesn't really seem useful. I'd simply remove it and move its content in the `init` function

File: src/coffee/run.coffee [hunk 14]
```diff
@@ -54,6 +75,50 @@ zipFiles = (path, filename) ->
...
+            Promise.resolve()
+      .catch (e) ->
+        console.error "Oops, there was an error reading the files in target folder #{argv.target}: #{e.message}"
+        process.exit 1
+
+importSphereProductTypes = (data) ->
+  importer = new ProductTypeImporter
+  importer.init(argv)
+  .then ->
+    importer.import data
```
Comment (emmenko) [inline src/coffee/run.coffee]: Looks like you can simply do

```coffeescript
importer = new ProductTypeImporter(argv)
importer.run(data)
```

File: src/coffee/product-type-import.coffee [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+ProductTypeImport = require 'sphere-product-type-import'
+{ExtendedLogger, ProjectCredentialsConfig} = require 'sphere-node-utils'
+package_json = require '../package.json'
+
+class ProductTypeImporter
+
+  ###
+  Prepare for importing, initialize used modules
+  ###
+  init: (argv) ->
```
Comment (emmenko) [inline src/coffee/product-type-import.coffee]: This should be the `constructor`

Comment (junajan) [inline src/coffee/product-type-import.coffee]: I had problems with constructor containing asynchronous operation

Comment (emmenko) [inline src/coffee/product-type-import.coffee]: Well, the async stuff shouldn't go in the contructor :)

I think just a function `run` or `import` is enough, you just do everything there.

File: src/coffee/product-type-import.coffee [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    @_ensureCredentials argv
+    .then (credentials) =>
+      @sphereImporter = new ProductTypeImport.default(logger, credentials)
+
+  ###
+  Import product types using sphere product import tool
+  ###
+  import: (data) ->
+    console.log "Importing product types to sphere"
+    Promise.map data.productTypes, (productType) ->
```
Comment (emmenko) [inline src/coffee/product-type-import.coffee]: Might consider to put a concurrency limit here

File: src/coffee/product-type-import.coffee (hunk 15 above)
Comment (junajan) [inline src/coffee/product-type-import.coffee]: but than it would have to create ProductTypeImport object in that import function which is wrong because it should be initialized on some central place so it can be possibly used on other places

File: features/common.feature (hunk 7 above)
Comment (junajan) [inline features/common.feature]: because travis had configured env variables for nicola project - thats why there were failing tests

File: package.json (hunk 8 above)
Comment (junajan) [inline package.json]: updated 4866775

File: package.json (hunk 9 above)
Comment (junajan) [inline package.json]: updated 4866775

File: src/coffee/product-type-import.coffee (hunk 10 above)
Comment (junajan) [inline src/coffee/product-type-import.coffee]: updated 4866775

File: src/coffee/product-type-import.coffee (hunk 11 above)
Comment (junajan) [inline src/coffee/product-type-import.coffee]: updated 4866775

File: src/coffee/run.coffee (hunk 12 above)
Comment (junajan) [inline src/coffee/run.coffee]: updated 4866775

File: src/coffee/product-type-import.coffee (hunk 13 above)
Comment (junajan) [inline src/coffee/product-type-import.coffee]: updated 4866775

File: src/coffee/run.coffee (hunk 14 above)
Comment (junajan) [inline src/coffee/run.coffee]: Not sure how to initialize async structures in constructor

File: src/coffee/product-type-import.coffee (hunk 16 above)
Comment (junajan) [inline src/coffee/product-type-import.coffee]: added here 4866775

File: src/coffee/run.coffee (hunk 14 above)
Comment (emmenko) [inline src/coffee/run.coffee]: What is exactly the async stuff that we're talking about here?

Comment (junajan) [inline src/coffee/run.coffee]: @emmenko - please check update here 631bdd8 - I added singleton function which will wrap init together with constructor..

That init function returns promise because it uses `ProjectCredentialsConfig.create()` in module `sphere-node-utils` which does some async stuff (like reading config file I guess)

File: src/coffee/product-type-import.coffee (hunk 15 above)
Comment (junajan) [inline src/coffee/product-type-import.coffee]: updated here 631bdd8

File: src/coffee/product-type-import.coffee [hunk 17]
```diff
@@ -78,4 +78,13 @@ class ProductTypeImporter
...
+    
+instance = null
+module.exports = (argv) ->
+  if not instance
+    instance = new ProductTypeImporter
+    instance.init(argv)
+    .then ->
+      instance
+  else
+    Promise.resolve instance
```
Comment (emmenko) [inline src/coffee/product-type-import.coffee]: This looks more complicated. Leave it as it was before then.

Comment (junajan) [inline src/coffee/product-type-import.coffee]: oki.. reverted last commit

File: src/spec/product-type-importer.spec.coffee [hunk 18]
```diff
@@ -0,0 +1,113 @@
+Promise = require 'bluebird'
+ProductTypeImporter = require '../lib/product-type-import'
+{SphereClient} = require 'sphere-node-sdk'
+
+errMissingCredentials = 'Missing configuration in env variables'
+
+argv =
+  clientId: process.env.SPHERE_CLIENT_ID
+  clientSecret: process.env.SPHERE_CLIENT_SECRET
+  projectKey: process.env.SPHERE_PROJECT_KEY
```
Comment (emmenko) [inline src/spec/product-type-importer.spec.coffee]: Usually we put credentials into a file and read it from there

https://github.com/sphereio/sphere-node-sdk/blob/master/create_config.sh

So we can also run the tests locally without env vars

File: src/spec/product-type-importer.spec.coffee [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+      },
+      "attributeConstraint": "None",
+      "isRequired": false,
+      "isSearchable": false,
+      "inputHint": "MultiLine"
+    }
+  ]
+}
+
+describe 'ProductTypeImporter', ->
```
Comment (emmenko) [inline src/spec/product-type-importer.spec.coffee]: I guess those are integration tests. Can you put this file into an `spec/integration` folder?

File: src/spec/product-type-importer.spec.coffee [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+    importer.init(argv)
+    .then ->
+      sphereClient.productTypes.fetch()
+      .then (res) ->
+        console.log "Deleting old product types", res.body.results.length
+        Promise.map res.body.results, (productType) ->
+          sphereClient.productTypes.byId(productType.id)
+          .delete(productType.version)
+        .then ->
+          done()
```
Comment (emmenko) [inline src/spec/product-type-importer.spec.coffee]: Nitpick: you can keep the chain as a flat structure

```coffeescript
importer.init(argv)
.then -> sphereClient.productTypes.fetch()
.then (res) ->
  console.log "Deleting old product types", res.body.results.length
  Promise.map res.body.results, (productType) ->
    sphereClient.productTypes.byId(productType.id)
    .delete(productType.version)
.then done
```

File: src/spec/product-type-importer.spec.coffee [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+        'client_secret': argv.clientSecret
+        'project_key': argv.projectKey
+      stats:
+        includeHeaders: true
+        maskSensitiveHeaderData: true
+    timeout: 360000
+
+    sphereClient = new SphereClient config
+    importer = new ProductTypeImporter
+    importer.init(argv)
```
Comment (emmenko) [inline src/spec/product-type-importer.spec.coffee]: Why are you importing stuff `beforeEach` test?

Comment (emmenko) [inline src/spec/product-type-importer.spec.coffee]: Usually we use `beforeEach` for cleaning up the data

File: src/spec/product-type-importer.spec.coffee (hunk 18 above)
Comment (junajan) [inline src/spec/product-type-importer.spec.coffee]: Updated here a4f15b6 - now we take only project_key from env and the rest can be saved in some file (eg. config.json)

Comment (butenkor) [inline src/spec/product-type-importer.spec.coffee]: @emmenko we used to do that but it is not very useful if you run it for example in iron.io. as you can set env vars but can not generate files on file system before the job runs. I would rather start using env vars.

Comment (emmenko) [inline src/spec/product-type-importer.spec.coffee]: But you don't run tests on iron.io. I'm just talking about tests here. Running locally like this it's not very helpful.

Comment (butenkor) [inline src/spec/product-type-importer.spec.coffee]: Sure i would leave it as an option too but env vars are also required for CI, right?

File: src/spec/product-type-importer.spec.coffee (hunk 21 above)
Comment (junajan) [inline src/spec/product-type-importer.spec.coffee]: I wanted to use `before` or `beforeAll` but it looks like that jasmine-node does not have this function yet (it uses old version of jasmine). I will try to switch it to mocha on Monday.

File: src/spec/product-type-importer.spec.coffee (hunk 20 above)
Comment (junajan) [inline src/spec/product-type-importer.spec.coffee]: updated here 48a32c2

File: src/spec/product-type-importer.spec.coffee (hunk 19 above)
Comment (junajan) [inline src/spec/product-type-importer.spec.coffee]: updated here 48a32c2

File: src/spec/product-type-importer.spec.coffee (hunk 21 above)
Comment (junajan) [inline src/spec/product-type-importer.spec.coffee]: Switched to mocha in 48a32c2 - it has native support for beforeAll/afterAll functions

File: src/spec/integration/product-type-importer.spec.coffee [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+      "inputHint": "MultiLine"
+    }
+  ]
+}
+
+describe 'ProductTypeImporter', ->
+  importer = null
+  sphereClient = null
+
+  before ->
```
Comment (emmenko) [inline src/spec/integration/product-type-importer.spec.coffee]: You're doing async stuff here, you have to use the `done` callback.

File: src/spec/integration/product-type-importer.spec.coffee [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+        clientSecret: sphereCredentials.client_secret
+
+      importer = new ProductTypeImporter
+      importer.init config
+
+      options =
+        config: sphereCredentials
+      sphereClient = new SphereClient options
+
+  beforeEach ->
```
Comment (emmenko) [inline src/spec/integration/product-type-importer.spec.coffee]: Same here with the `done` callback

File: src/spec/integration/product-type-importer.spec.coffee [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+
+describe 'ProductTypeImporter', ->
+  importer = null
+  sphereClient = null
+
+  before ->
+    expect(argv.projectKey).to.be.a 'string', errMissingCredentials
+
+    ProjectCredentialsConfig.create()
+    .then (credentials) ->
```
Comment (emmenko) [inline src/spec/integration/product-type-importer.spec.coffee]: If you want to keep this, please document in the readme that you need to have a file according to the credentials setup

File: src/spec/integration/product-type-importer.spec.coffee [hunk 25]
```diff
@@ -0,0 +1,108 @@
+{expect} = require 'chai'
+Promise = require 'bluebird'
+ProductTypeImporter = require '../../lib/product-type-import'
+{SphereClient} = require 'sphere-node-sdk'
+{ProjectCredentialsConfig} = require 'sphere-node-utils'
+
+errMissingCredentials = 'Missing configuration in env variables'
+
+argv =
+  projectKey: process.env.SPHERE_PROJECT_KEY
```
Comment (emmenko) [inline src/spec/integration/product-type-importer.spec.coffee]: I still find this really unusable for local development.

File: src/spec/integration/product-type-importer.spec.coffee [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+    expect(importer).to.be.an 'object'
+    expect(sphereClient).to.be.an 'object'
+
+    sphereClient.productTypes.fetch()
+    .then (res) ->
+      expect(res.body.results.length).to.equal 0
+      console.log "Importing product type using importer"
+      importer.import {productTypes: [testProductType]}
+    .then ->
+      console.log "Product type imported - verifiing result"
```
Comment (emmenko) [inline src/spec/integration/product-type-importer.spec.coffee]: Remove

File: src/spec/integration/product-type-importer.spec.coffee [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+      sphereClient.productTypes.fetch()
+    .then (res) ->
+      expect(res.body.results.length).to.equal 1
+
+  it 'should not import wrong product type', (done) ->
+    expect(importer).to.be.an 'object'
+    expect(sphereClient).to.be.an 'object'
```
Comment (emmenko) [inline src/spec/integration/product-type-importer.spec.coffee]: You don't need those assertions

File: src/spec/integration/product-type-importer.spec.coffee [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+
+    delete testProductType.name
```
Comment (emmenko) [inline src/spec/integration/product-type-importer.spec.coffee]: You're mutating the object, make a shallow copy first.

File: src/spec/integration/product-type-importer.spec.coffee [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+    importer.import {productTypes: [testProductType]}
+    .then ->
+      done "Importer wrong product type"
+    .catch ->
+      sphereClient.productTypes.fetch()
+      .then (res) ->
+        expect(res.body.results.length).to.equal 0
+        done()
+    return 0 # do not return promise
```
Comment (emmenko) [inline src/spec/integration/product-type-importer.spec.coffee]: This is not necessary

File: src/spec/integration/product-type-importer.spec.coffee [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+
+  it 'should not import wrong product type', (done) ->
+    expect(importer).to.be.an 'object'
+    expect(sphereClient).to.be.an 'object'
+
+    delete testProductType.name
+    importer.import {productTypes: [testProductType]}
+    .then ->
+      done "Importer wrong product type"
+    .catch ->
```
Comment (emmenko) [inline src/spec/integration/product-type-importer.spec.coffee]: I would probably check here that the error is correct

File: src/spec/integration/product-type-importer.spec.coffee (hunk 22 above)
Comment (junajan) [inline src/spec/integration/product-type-importer.spec.coffee]: I switched it to mocha which is working with Promises too - so it uses promise instead of done callback

File: src/spec/integration/product-type-importer.spec.coffee (hunk 23 above)
Comment (junajan) [inline src/spec/integration/product-type-importer.spec.coffee]: it also uses returned promise instead of callback

File: src/spec/integration/product-type-importer.spec.coffee (hunk 25 above)
Comment (junajan) [inline src/spec/integration/product-type-importer.spec.coffee]: It takes credentials from .sphere config file now. How would you want to pass project-key to tests?

File: src/spec/integration/product-type-importer.spec.coffee (hunk 30 above)
Comment (junajan) [inline src/spec/integration/product-type-importer.spec.coffee]: updated here 0f5fb54

File: src/spec/integration/product-type-importer.spec.coffee (hunk 29 above)
Comment (junajan) [inline src/spec/integration/product-type-importer.spec.coffee]: return is important here because coffeescript returns last variable as a function result and Mocha does not work when it receives Promise and called callback function at the same time.

File: src/spec/integration/product-type-importer.spec.coffee (hunk 26 above)
Comment (junajan) [inline src/spec/integration/product-type-importer.spec.coffee]: Why removing logs? They can be useful later..

File: src/spec/integration/product-type-importer.spec.coffee (hunk 24 above)
Comment (junajan) [inline src/spec/integration/product-type-importer.spec.coffee]: I will add few lines to the documentation how to run tests

Comment (junajan) [inline src/spec/integration/product-type-importer.spec.coffee]: added here 5fcb461

File: src/spec/integration/product-type-importer.spec.coffee (hunk 25 above)
Comment (junajan) [inline src/spec/integration/product-type-importer.spec.coffee]: added documentation how to start tests here 5fcb461

File: src/coffee/product-type-import.coffee [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+  ###
+  Prepare for importing, initialize used modules
+  ###
+  init: (argv) ->
+    @logOptions =
+      name: "#{package_json.name}-#{package_json.version}"
+      silent: !! argv.logSilent
+      streams: [
+        { level: 'error', stream: process.stderr }
+        { level: argv.logLevel || 'info', path: "#{argv.logDir || '.'}/#{package_json.name}.log" }
```
Comment (PhilippSpo) [inline src/coffee/product-type-import.coffee]: It might be better to use `path.join` here

File: src/spec/integration/product-type-importer.spec.coffee (hunk 26 above)
Comment (PhilippSpo) [inline src/spec/integration/product-type-importer.spec.coffee]: Since it is in the tests, I don't think it will be useful later ..

Comment (junajan) [inline src/spec/integration/product-type-importer.spec.coffee]: removed here e866a9d

File: src/coffee/product-type-import.coffee (hunk 31 above)
Comment (junajan) [inline src/coffee/product-type-import.coffee]: added here 09fdc1b

**Comment by PhilippSpo** [comment]
OK LGTM now 👍
